llvm/lib/Transforms/Scalar/LoopLoadElimination.cpp

```diff
@@ -462,14 +466,20 @@ class LoadEliminationForLoop {
            "The type sizes should match!");
 
     Value *StoreValue = Cand.Store->getValueOperand();
-    if (LoadType != StoreType)
+    if (LoadType != StoreType) {
       StoreValue = CastInst::CreateBitOrPointerCast(StoreValue, LoadType,
                                                     "store_forward_cast",
                                                     Cand.Store->getIterator());
+      // Because it casts the old `load` value and is used by the new `phi`
+      // which replaces the old `load`, we give the `load`'s debug location
+      // to it.
+      cast<Instruction>(StoreValue)->setDebugLoc(Cand.Load->getDebugLoc());
+    }
 
     PHI->addIncoming(StoreValue, L->getLoopLatch());
 
     Cand.Load->replaceAllUsesWith(PHI);
+    PHI->setDebugLoc(Cand.Load->getDebugLoc());
```

Comment:
This is an interesting case. When promoting store/loads I think we usually give phis a debugloc from the store (or if it's from stores, plural, it's a merged location). At least that's what we do in mem2reg. I think in this case using the `load`'s DebugLoc makes intuitive sense - we're just eliminating the load here after all, but I'd be interested to hear what others think about this case too (cc @SLTozer , @adrian-prantl , @dwblaikie, @jryans - there's a comment at the top of the function that explains the transformation) and what you make of this @Apochens?

Reply:
Using the load's location here does make intuitive sense to me as well. I'm less sure what we've done for such cases historically though, so I'm interested to hear what others think as well.

Reply:
> I think in this case using the load's DebugLoc makes intuitive sense - we're just eliminating the load here after all

That's what I agree with. 

> When promoting store/loads I think we usually give phis a debugloc from the store (or if it's from stores, plural, it's a merged location).

I'd like to know why give the store'd debug location to an instruction that replaces a load instruction. Are there some conventions or something?

Reply:
Hmm, should we ping other guys again? 🤔

Reply:
Probably best to wait a few more days first, as people may be on holiday or have a long queue to work through.

Reply:
Seems like enough of us have appeared and all concluded the same thing here, so I think we can leave this part as it is.